server/src/main/java/org/elasticsearch/index/store/Store.java

```diff
@@ -786,26 +778,21 @@ public MetadataSnapshot(Map<String, StoreFileMetadata> metadata, Map<String, Str
             assert metadata.isEmpty() || numSegmentFiles() == 1 : "numSegmentFiles: " + numSegmentFiles();
         }
 
-        /**
-         * Read from a stream.
-         */
-        public MetadataSnapshot(StreamInput in) throws IOException {
-            final int size = in.readVInt();
-            Map<String, StoreFileMetadata> metadata = new HashMap<>();
-            for (int i = 0; i < size; i++) {
-                StoreFileMetadata meta = new StoreFileMetadata(in);
-                metadata.put(meta.name(), meta);
-            }
-            Map<String, String> commitUserData = new HashMap<>();
-            int num = in.readVInt();
-            for (int i = num; i > 0; i--) {
-                commitUserData.put(in.readString(), in.readString());
+        public static MetadataSnapshot readFrom(StreamInput in) throws IOException {
+            final int metadataSize = in.readVInt();
+            final Map<String, StoreFileMetadata> metadata = metadataSize == 0 ? emptyMap() : new HashMap<>();
```

Comment:
NIT: we might use `org.elasticsearch.common.util.Maps#newMapWithExpectedSize` to avoid resizing the map

Reply:
Also may be it is worth generalizing this pattern of reading with something like:
`Map<K, V> readMap(Writeable.Reader<V> valueReader, Function<V, K> keyCreator) throws IOException` that would internally handle map creation, initialization and sizing?

Reply:
++ see 488a05ba906784121814fa5a8c395272397ed842

Looks like we are not properly sizing the map produced by `StreamInput#readMap` and `StreamInput#readOrderedMap` either. I'm not touching that in this PR, nor the map-from-list thing, but seems reasonable I think.

Reply:
https://github.com/elastic/elasticsearch/pull/84045 :)

Reply:
I could also open a followup pr on map-from-list